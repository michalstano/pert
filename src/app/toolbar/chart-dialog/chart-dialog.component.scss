:host {
  display: grid;
  width: 100%;
  height: 100%;
  grid-template-rows: 1fr auto;
}

.mat-dialog-content {
  overflow: hidden;
  padding: 0;
  margin: 0 1rem;
}
