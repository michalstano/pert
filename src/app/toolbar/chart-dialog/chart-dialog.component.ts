import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
import { LegendPosition } from '@swimlane/ngx-charts';
import { ChartItem } from '../../sandbox/+state/sandbox.model';

@Component({
  selector: 'app-chart-dialog',
  template: `
    <div mat-dialog-content>
      <ngx-charts-bar-horizontal-stacked
        [scheme]="colorScheme"
        [results]="data.results"
        [gradient]="false"
        [xAxis]="true"
        [yAxis]="true"
        [legend]="false"
        [showXAxisLabel]="true"
        [showYAxisLabel]="false"
        [xAxisLabel]="'Czas'"
        [yAxisLabel]="'Zadania'"
        [xScaleMax]="10"
        [tooltipDisabled]="true"
      >
      </ngx-charts-bar-horizontal-stacked>
    </div>
    <div mat-dialog-actions>
      <button mat-button (click)="dismiss()">
        Zamknij
      </button>
    </div>
  `,
  styleUrls: ['./chart-dialog.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ChartDialogComponent {
  LegendPosition = LegendPosition;

  readonly colorScheme = {
    domain: ['transparent', 'var(--primary-color)', 'var(--secondary-color)']
  };

  constructor(
    readonly dialogRef: MatDialogRef<ChartDialogComponent>,
    @Inject(MAT_DIALOG_DATA) readonly data: { results: ChartItem[] }
  ) {}

  dismiss(): void {
    this.dialogRef.close();
  }
}
