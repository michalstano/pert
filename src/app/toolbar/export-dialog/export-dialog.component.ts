import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { Validators } from '@angular/forms';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
import { FormControl } from '@ngneat/reactive-forms';

export interface DialogData {
  name: string;
}

@Component({
  selector: 'app-export-dialog',
  template: `
    <div mat-dialog-content>
      <mat-icon>file_upload</mat-icon>
      <p>
        Czy na pewno chcesz zapisać dane do pliku? Jeśli tak, to podaj nazwę
        pliku.
      </p>
      <mat-form-field class="example-full-width" appearance="fill">
        <mat-label>Nazwa pliku</mat-label>
        <input
          type="text"
          matInput
          cdkFocusInitial
          max-length="10"
          [formControl]="control"
        />
      </mat-form-field>
    </div>
    <div mat-dialog-actions>
      <button mat-button (click)="dismiss()">
        Anuluj
      </button>
      <button
        mat-button
        class="accept-btn"
        [mat-dialog-close]="control.value"
        [disabled]="control.invalid"
      >
        Zapisz
      </button>
    </div>
  `,
  styleUrls: ['./export-dialog.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ExportDialogComponent {
  readonly control = new FormControl<string>('', {
    validators: [Validators.required]
  });

  constructor(
    readonly dialogRef: MatDialogRef<ExportDialogComponent>,
    @Inject(MAT_DIALOG_DATA) readonly data: DialogData
  ) {}

  dismiss(): void {
    this.dialogRef.close();
  }
}
