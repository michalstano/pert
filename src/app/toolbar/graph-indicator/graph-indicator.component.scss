:host {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: black;
  margin-bottom: 1rem;
}

p {
  margin: 0.3rem 0;
  font-size: 0.85rem;
  color: var(--text);
}

mat-icon {
  height: 3rem;
  width: 3rem;
  font-size: 3rem;

  &.correct-icon {
    color: green;
  }

  &.incorrect-icon {
    color: red;
  }
}
