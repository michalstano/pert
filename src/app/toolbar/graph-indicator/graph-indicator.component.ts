import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

@Component({
  selector: 'app-graph-indicator',
  template: `
    <ng-container [ngSwitch]="isGraphCorrect">
      <ng-container *ngSwitchCase="true">
        <mat-icon class="correct-icon">check</mat-icon>
        <p>Poprawny</p>
      </ng-container>
      <ng-container *ngSwitchCase="false">
        <mat-icon class="incorrect-icon">clear</mat-icon>
        <p>Niepoprawny</p>
      </ng-container>
    </ng-container>
  `,
  styleUrls: ['./graph-indicator.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class GraphIndicatorComponent {
  @Input() isGraphCorrect: boolean | undefined = undefined;
}
