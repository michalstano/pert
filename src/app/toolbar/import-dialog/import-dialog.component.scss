:host {
  color: var(--text);
}

mat-icon {
  width: 100%;
  height: 100%;
  font-size: 60px;
  text-align: center;
}

p {
  margin: 0 0 1rem 0;
}

button {
  font-weight: bold;
}

.accept-btn {
  color: green;
}
