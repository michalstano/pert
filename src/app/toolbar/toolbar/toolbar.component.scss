:host {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--light-background2);
  color: var(--secondary-color);
  border-right: 1px solid var(--light-border-color);
  height: 100%;
}

.agh-logo {
  width: 5rem;
  height: 5rem;
  padding: 1rem 0;
}

.line {
  display: block;
  width: 80%;
  height: 1px;
  background-color: var(--light-border-color);
}
