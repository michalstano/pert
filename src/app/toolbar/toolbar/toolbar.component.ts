import { ChangeDetectionStrategy, Component } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
import { Store } from '@ngrx/store';
import { HotToastService } from '@ngneat/hot-toast';
import { take } from 'rxjs/operators';
import { SandboxFacade } from '../../sandbox/+state/sandbox.facade';
import { ToolbarActions } from '../+state/toolbar.actions';
import { ExportDialogComponent } from '../export-dialog/export-dialog.component';
import { ImportDialogComponent } from '../import-dialog/import-dialog.component';
import { PortData } from '../../sandbox/+state/sandbox.model';
import { ChartDialogComponent } from '../chart-dialog/chart-dialog.component';
import { InfoWindowComponent } from '../info-window/info-window.component';

@UntilDestroy()
@Component({
  selector: 'app-toolbar',
  template: `
    <svg-icon key="agh-logo" class="agh-logo"></svg-icon>
    <div class="line"></div>
    <button
      mat-icon-button
      matTooltip="Stwórz węzeł"
      matTooltipPosition="after"
      (click)="dispatchAddAoNButtonClicked()"
      data-test="create-node-btn"
    >
      <svg-icon key="add-aon-block"></svg-icon>
    </button>
    <button
      mat-icon-button
      matTooltip="Usuń węzeł/połączenie"
      matTooltipPosition="after"
      [disabled]="(sandboxFacade.isPossibleToRemoveItem$ | async) === false"
      (click)="dispatchRemoveItemButtonClicked()"
      data-test="remove-btn"
    >
      <mat-icon>delete</mat-icon>
    </button>
    <button
      mat-icon-button
      matTooltip="Połącz węzły"
      matTooltipPosition="after"
      [disabled]="!(sandboxFacade.isPossibleToEnableConnectionMode$ | async)"
      (click)="dispatchConnectNodesButtonClicked()"
      data-test="connect-nodes-btn"
    >
      <mat-icon>compare_arrows</mat-icon>
    </button>
    <div class="line"></div>
    <button
      mat-icon-button
      matTooltip="Importuj dane"
      matTooltipPosition="after"
      (click)="openImportDialog()"
      data-test="import-btn"
    >
      <mat-icon>file_download</mat-icon>
    </button>
    <button
      mat-icon-button
      matTooltip="Eksportuj dane"
      matTooltipPosition="after"
      [disabled]="!(sandboxFacade.isPortData$ | async)"
      (click)="openExportDialog()"
      data-test="export-btn"
    >
      <mat-icon>file_upload</mat-icon>
    </button>
    <div class="line"></div>
    <button
      mat-icon-button
      matTooltip="Informacje"
      matTooltipPosition="after"
      (click)="openInfo()"
      data-test="info-btn"
    >
      <mat-icon>info</mat-icon>
    </button>
    <div
      *ngIf="(sandboxFacade.isGraphCorrect$ | async) !== undefined"
      class="down"
    >
      <button
        *ngIf="sandboxFacade.isGraphCorrect$ | async"
        mat-icon-button
        matTooltip="Podejrzyj wykres Gantta"
        matTooltipPosition="after"
        (click)="generateChartButtonClicked()"
        data-test="chart-btn"
      >
        <mat-icon>bar_chart</mat-icon>
      </button>
      <app-graph-indicator
        [isGraphCorrect]="sandboxFacade.isGraphCorrect$ | async"
      ></app-graph-indicator>
    </div>
  `,
  styleUrls: ['./toolbar.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ToolbarComponent {
  constructor(
    private readonly store: Store,
    private readonly dialog: MatDialog,
    readonly sandboxFacade: SandboxFacade,
    private readonly toast: HotToastService
  ) {}

  dispatchAddAoNButtonClicked(): void {
    this.store.dispatch(ToolbarActions.addAoNButtonClicked());
  }

  dispatchRemoveItemButtonClicked(): void {
    this.store.dispatch(ToolbarActions.removeItemButtonClicked());
  }

  dispatchConnectNodesButtonClicked(): void {
    this.store.dispatch(ToolbarActions.connectNodesButtonClicked());
  }

  openExportDialog(): void {
    const dialogRef = this.dialog.open(ExportDialogComponent, {
      panelClass: 'dialog'
    });
    dialogRef
      .afterClosed()
      .pipe(untilDestroyed(this))
      .subscribe((result: string | undefined) => {
        if (result) {
          this.store.dispatch(
            ToolbarActions.exportButtonClicked({ fileName: result })
          );
        }
      });
  }

  openImportDialog(): void {
    const dialogRef = this.dialog.open(ImportDialogComponent, {
      panelClass: 'dialog'
    });
    dialogRef
      .afterClosed()
      .pipe(untilDestroyed(this))
      .subscribe((result: PortData) => {
        if (result) {
          this.store.dispatch(ToolbarActions.importButtonClicked({ result }));
        }
      });
  }

  openInfo(): void {
    this.toast.show(InfoWindowComponent, {
      id: 'info-window',
      dismissible: true,
      autoClose: false,
      position: 'bottom-right',
      style: {
        border: '1px solid var(--light-border-color)',
        padding: '16px',
        ['background-color']: 'var(--light-background2)',
        color: 'var(--text)'
      }
    });
  }

  generateChartButtonClicked(): void {
    this.sandboxFacade.chartItems$.pipe(take(1)).subscribe(chartItems =>
      this.dialog.open(ChartDialogComponent, {
        panelClass: 'dialog',
        width: '50%',
        height: '50%',
        data: {
          results: chartItems
        }
      })
    );
  }
}
