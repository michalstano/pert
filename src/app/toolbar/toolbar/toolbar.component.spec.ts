import { MatDialog } from '@angular/material/dialog';
import { MatIcon } from '@angular/material/icon';
import { createComponentFactory, Spectator } from '@ngneat/spectator';
import { Store } from '@ngrx/store';
import { SvgIconComponent } from '@ngneat/svg-icon';
import { HotToastService } from '@ngneat/hot-toast';
import { of } from 'rxjs';
import { MockComponents } from 'ng-mocks';
import { SandboxFacade } from '../../sandbox/+state/sandbox.facade';
import { GraphIndicatorComponent } from '../graph-indicator/graph-indicator.component';
import { ToolbarComponent } from './toolbar.component';
import { ToolbarActions } from '../+state/toolbar.actions';
import { ImportDialogComponent } from '../import-dialog/import-dialog.component';
import { ExportDialogComponent } from '../export-dialog/export-dialog.component';
import { ChartDialogComponent } from '../chart-dialog/chart-dialog.component';
import { createChartItem } from '../../sandbox/+state/sandbox.model.mock';

describe('Toolbar', () => {
  let spectator: Spectator<ToolbarComponent>;
  let component: ToolbarComponent;
  let store: Store;
  let sandboxFacade: SandboxFacade;
  let dialog: MatDialog;
  let toast: HotToastService;

  const mocks = {
    chartItems: [createChartItem()]
  };

  const createComponent = createComponentFactory({
    component: ToolbarComponent,
    providers: [
      {
        provide: Store,
        useValue: {
          dispatch: jest.fn()
        }
      },
      {
        provide: MatDialog,
        useValue: {
          open: jest.fn(() => ({
            afterClosed: () => of(true)
          }))
        }
      },
      {
        provide: SandboxFacade,
        useValue: {
          isPossibleToRemoveItem$: of(true),
          isPossibleToEnableConnectionMode$: of(true),
          isPortData$: of(true),
          isGraphCorrect$: of(true),
          chartItems$: of(mocks.chartItems)
        }
      },
      {
        provide: HotToastService,
        useValue: {
          show: jest.fn()
        }
      }
    ],
    declarations: [
      ...MockComponents(GraphIndicatorComponent, MatIcon, SvgIconComponent)
    ]
  });

  const selectors = {
    createNodeBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="create-node-btn"]'),
    removeBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="remove-btn"]'),
    connectNodesBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="connect-nodes-btn"]'),
    importBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="import-btn"]'),
    exportBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="export-btn"]'),
    infoBtn: () => spectator.query<HTMLButtonElement>('[data-test="info-btn"]'),
    chartBtn: () =>
      spectator.query<HTMLButtonElement>('[data-test="chart-btn"]')
  };

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
    store = spectator.inject(Store);
    sandboxFacade = spectator.inject(SandboxFacade);
    dialog = spectator.inject(MatDialog);
    toast = spectator.inject(HotToastService);
  });

  it('should match snapshot', () => {
    expect(spectator.fixture).toMatchSnapshot();
  });

  describe('create-node-btn', () => {
    it('on (click) should dispatch AddAoNButtonClicked action', () => {
      selectors.createNodeBtn().click();

      expect(store.dispatch).toHaveBeenCalledWith(
        ToolbarActions.addAoNButtonClicked()
      );
    });
  });

  describe('remove-btn', () => {
    it('on (click) should dispatch removeItemButtonClicked action', () => {
      selectors.removeBtn().click();

      expect(store.dispatch).toHaveBeenCalledWith(
        ToolbarActions.removeItemButtonClicked()
      );
    });

    describe('[disabled]', () => {
      it('should be set to true if there is no possible to remove item', () => {
        sandboxFacade.isPossibleToRemoveItem$ = of(false);
        spectator.detectComponentChanges();

        expect(selectors.removeBtn().disabled).toBeTruthy();
      });

      it('should be set to false if there is possible to remove item', () => {
        sandboxFacade.isPossibleToRemoveItem$ = of(true);
        spectator.detectComponentChanges();

        expect(selectors.removeBtn().disabled).toBeFalsy();
      });
    });
  });

  describe('connect-nodes-btn', () => {
    it('on (click) should dispatch connectNodesButtonClicked action', () => {
      selectors.connectNodesBtn().click();

      expect(store.dispatch).toHaveBeenCalledWith(
        ToolbarActions.connectNodesButtonClicked()
      );
    });

    describe('[disabled]', () => {
      it('should be set to true if there is no possible to enable connection mode', () => {
        sandboxFacade.isPossibleToEnableConnectionMode$ = of(false);
        spectator.detectComponentChanges();

        expect(selectors.connectNodesBtn().disabled).toBeTruthy();
      });

      it('should be set to false if there is possible to enable connectionMode', () => {
        sandboxFacade.isPossibleToEnableConnectionMode$ = of(true);
        spectator.detectComponentChanges();

        expect(selectors.connectNodesBtn().disabled).toBeFalsy();
      });
    });
  });

  describe('import-btn', () => {
    it('on (click) should open import dialog', () => {
      selectors.importBtn().click();

      expect(dialog.open).toHaveBeenCalledWith(ImportDialogComponent, {
        panelClass: 'dialog'
      });
    });
  });

  describe('export-btn', () => {
    it('on (click) should open export dialog', () => {
      selectors.exportBtn().click();

      expect(dialog.open).toHaveBeenCalledWith(ExportDialogComponent, {
        panelClass: 'dialog'
      });
    });

    describe('[disabled]', () => {
      it('should be set to true if there is no port data', () => {
        sandboxFacade.isPortData$ = of(false);
        spectator.detectComponentChanges();

        expect(selectors.exportBtn().disabled).toBeTruthy();
      });

      it('should be set to false if there is port data', () => {
        sandboxFacade.isPortData$ = of(true);
        spectator.detectComponentChanges();

        expect(selectors.exportBtn().disabled).toBeFalsy();
      });
    });
  });

  describe('info-btn', () => {
    it('on (click) should open toast', () => {
      selectors.infoBtn().click();

      expect(toast.show).toHaveBeenCalled();
    });
  });

  describe('chart-btn', () => {
    it('should be visible if graph is correct', () => {
      sandboxFacade.isGraphCorrect$ = of(true);
      spectator.detectComponentChanges();

      expect(selectors.chartBtn()).toBeTruthy();
    });

    it('should not be visible if graph is not correct', () => {
      sandboxFacade.isGraphCorrect$ = of(false);
      spectator.detectComponentChanges();

      expect(selectors.chartBtn()).toBeFalsy();
    });

    it('on (click) should open chart dialog', () => {
      sandboxFacade.isGraphCorrect$ = of(true);
      spectator.detectComponentChanges();

      selectors.chartBtn().click();

      expect(dialog.open).toHaveBeenCalledWith(ChartDialogComponent, {
        panelClass: 'dialog',
        width: '50%',
        height: '50%',
        data: {
          results: mocks.chartItems
        }
      });
    });
  });
});
