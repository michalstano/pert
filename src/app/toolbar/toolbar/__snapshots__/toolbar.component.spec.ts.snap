// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar should match snapshot 1`] = `
<app-toolbar
  __ngContext__={[Function LRootView]}
  dialog={[Function Object]}
  sandboxFacade={[Function Object]}
  store={[Function Object]}
  toast={[Function Object]}
>
  <svg-icon
    class="agh-logo"
    key="agh-logo"
  /><div
    class="line"
  /><button
    data-test="create-node-btn"
    mat-icon-button=""
    mattooltip="Stwórz węzeł"
    mattooltipposition="after"
  >
    <svg-icon
      key="add-aon-block"
    />
  </button><button
    data-test="remove-btn"
    mat-icon-button=""
    mattooltip="Usuń węzeł/połączenie"
    mattooltipposition="after"
  >
    <mat-icon>
      delete
    </mat-icon>
  </button><button
    data-test="connect-nodes-btn"
    mat-icon-button=""
    mattooltip="Połącz węzły"
    mattooltipposition="after"
  >
    <mat-icon>
      compare_arrows
    </mat-icon>
  </button><div
    class="line"
  /><button
    data-test="import-btn"
    mat-icon-button=""
    mattooltip="Importuj dane"
    mattooltipposition="after"
  >
    <mat-icon>
      file_download
    </mat-icon>
  </button><button
    data-test="export-btn"
    mat-icon-button=""
    mattooltip="Eksportuj dane"
    mattooltipposition="after"
  >
    <mat-icon>
      file_upload
    </mat-icon>
  </button><div
    class="line"
  /><button
    data-test="info-btn"
    mat-icon-button=""
    mattooltip="Informacje"
    mattooltipposition="after"
  >
    <mat-icon>
      info
    </mat-icon>
  </button><div
    class="down"
  >
    <button
      data-test="chart-btn"
      mat-icon-button=""
      mattooltip="Podejrzyj wykres Gantta"
      mattooltipposition="after"
    >
      <mat-icon>
        bar_chart
      </mat-icon>
    </button>
    <app-graph-indicator />
  </div>
</app-toolbar>
`;
